analytics.php

```diff
@@ -0,0 +1,231 @@
+    ];
+    $attentiontable->attributes['class'] = 'w-auto admintable generaltable mr-4';
+    $htmlparams = ['class' => 'text-center'];
+    $totalattendees = count($attendees);
+    $data = [
+        [
+            get_string('view_analytics_number_of_students_speaking', 'bigbluebuttonbn'),
+            $attentiontotals['talks'].' / '.$totalattendees
+        ],
+        [
+            get_string('view_analytics_number_of_students_messaging', 'bigbluebuttonbn'),
+            $attentiontotals['chats'].' / '.$totalattendees
+        ],
+        [
+            get_string('view_analytics_number_of_students_using_emojis', 'bigbluebuttonbn'),
+            $attentiontotals['emojis'].' / '.$totalattendees
+        ],
+        [
+            get_string('view_analytics_number_of_students_raising_hands', 'bigbluebuttonbn'),
+            $attentiontotals['raisehand'].' / '.$totalattendees
+        ],
+    ];
+    $attentiontable->data = $data;
+    $attentiontablehtml = html_writer::table($attentiontable);
+    $tables .= $attentiontablehtml;
+
+    // Files box.
+    $filestable = new html_table();
+    $filestable->head = [get_string('view_analytics_heading_files', 'bigbluebuttonbn')];
+    $filestable->attributes['class'] = 'w-auto admintable generaltable';
+    $htmlparams = ['class' => 'text-center'];
+    $data = [];
+    foreach ($meetingsummary->meta->files as $file) {
+        $data[] = [$file];
+    }
+    $filestable->data = $data;
+    $filestablehtml = html_writer::table($filestable);
+    $tables .= $filestablehtml;
+
+    $output .= '<div class="d-flex align-items-start">'.$tables.'</div>';
+}
+
+if (!$table->is_downloading()) {
+    echo $output;
+}
+
+// Modify data - add in any poll data.
+if (!empty($meetingsummary->meta->polls)) {
+    $totalattendees = count($table->rawdata);
+    foreach ($meetingsummary->meta->polls as $index => $poll) {
+        $totalpollvotes = count((array)$poll->votes);
+        $pollid = 'poll'.$poll->id;
+        $table->columns[$pollid] = count($table->columns);
+        $table->column_class[$pollid] = 'text-center';
+        $table->column_style[$pollid] = '';
+        $table->column_suppress[$pollid] = false;
+
+        $newheader = get_string('view_analytics_poll', 'bigbluebuttonbn')." ".($index + 1);
+        if (!$table->is_downloading()) {
+            $newheader .= \html_writer::tag('div', (
+                    $totalpollvotes." / ".$totalattendees." ".get_string('view_analytics_poll_responses', 'bigbluebuttonbn')
+                ), ['class' => 'small']);
+        }
+        $table->headers[] = $newheader;
+
+        foreach ($table->rawdata as $rawattendeedata) {
+            // Note: This may to be rechecked for guests who participate, as their ids might be empty.
+            $rawattendeedata->{$pollid} = $poll->votes->{$rawattendeedata->userid} ?? '-';
+        }
+    }
+}
+
+$table->build_table();
+$table->finish_output();
+
+if (!$table->is_downloading()) {
+    echo $OUTPUT->footer();
+}
```

Comment:
In general with this file, it ends up quite monolithic. Can the individual tables be moved into functions in a renderer, and then this page will just essentially be 3 function calls?

Reply:
Possibly, I'll fix up the other issues and then try and move it out.